File: 0203-remove-linked-list-elements/go/dummy.go
package dummy

// Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

func removeElements(head *ListNode, val int) *ListNode {
	dummy := new(ListNode) // 新建一个虚拟头节点
	dummy.Next = head      // 虚拟头节点指向head
	notEqual := dummy      // 新建一个游标
	for notEqual.Next != nil {
		if notEqual.Next.Val == val {
			notEqual.Next = notEqual.Next.Next
		} else {
			notEqual = notEqual.Next
		}
	}
	return dummy.Next
}

File: 0203-remove-linked-list-elements/java/Dummy.java
// Definition for singly-linked list.
public class ListNode {
  int val;
  ListNode next;

  ListNode() {}

  ListNode(int val) { this.val = val; }

  ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

class Solution {
  // Time: O(n), Space: O(1)
  public ListNode removeElements(ListNode head, int val) {
    ListNode dummy = new ListNode(0);
    dummy.next = head;
    ListNode notEqual = dummy;

    while (notEqual.next != null) {
      if (notEqual.next.val == val) notEqual.next = notEqual.next.next;
      else notEqual = notEqual.next;
    }
    return dummy.next;
  }
}